notebooks/iclr2022: Split plots.ipynb into separate notebooks

tomtseng opened the pull request:
Refactors `plots.ipynb` by splitting into separate notebooks.

* Move first two cells of `plots.ipynb` into `utils.py`
* Split notebook into separate files based on what they plot
  * There's some code duplication in `checkpoint-transfer.ipynb` and `elo-vs-visits.ipynb` but I didn't bother de-duplicating it since we don't expect to want to rerun those plots often

tomtseng commented on notebooks/notebooks/iclr2022/elo-by-visits-2.pgf:
```diff
@@ -892,15 +892,15 @@
 \pgfusepath{stroke,fill}%
```
this plot got regenerated because the experiment it depends on is still running and outputting new data

AdamGleave commented on notebooks/notebooks/iclr2022/win-rate-vs-training-steps.ipynb:
```diff
@@ -0,0 +1,366 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #1.    import matplotlib.pyplot as plt
</pre><p>The code in this particular notebook feels fundamental enough that we likely want to move it <em>outside</em> of a notebook into a Python module -- but that can be another PR.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/HumanCompatibleAI/KataGoVisualizer/pull/21/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

AdamGleave replied:
<pre class="ql-syntax" spellcheck="false">Line #1.    def get_victim_active_ranges(df):
</pre><p>Add return type annotation? <code>Mapping[str, int]</code> I think. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/HumanCompatibleAI/KataGoVisualizer/pull/21/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

AdamGleave replied:
<pre class="ql-syntax" spellcheck="false">Line #13.            df19[["victim_name_v2", "adv_steps"]].groupby("victim_name_v2").min().adv_steps
</pre><p><code>groupby</code> returns an object which you can call multiple times -- so can put the groupby on its own line assigned to a temporary variable <code>grp</code>  say and then call <code>grp.min().adv_steps</code> and <code>grp.max().adv_steps</code>.I think you might also be able to do <code>grp.adv_steps.min()</code> and this seems more natural to me (rather than taking the min of everything and then just picking one attribute from it) but not 100% that works.      </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/HumanCompatibleAI/KataGoVisualizer/pull/21/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

AdamGleave replied:
<pre class="ql-syntax" spellcheck="false">Line #33.    def plot_training(
</pre><p>parameter type annotations?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/HumanCompatibleAI/KataGoVisualizer/pull/21/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

AdamGleave replied:
<pre class="ql-syntax" spellcheck="false">Line #38.        plot_name,
</pre><p>Alternative: return the <code>plt.Figure</code> and make it the callees responsible to save it? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/HumanCompatibleAI/KataGoVisualizer/pull/21/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

AdamGleave replied:
<pre class="ql-syntax" spellcheck="false">Line #41.    ):
</pre><p>Add return type annotation <code>-&gt; None</code>?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/HumanCompatibleAI/KataGoVisualizer/pull/21/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

AdamGleave replied:
<pre class="ql-syntax" spellcheck="false">Line #54.        df19 = df.loc[df.board_size == 19]
</pre><p>Nit but inclined to make the board size a parameter, that defaults to 19, as we might in fact want to look at win rates of different board sizes in future. Also always nice to get rid of magic numbers in code.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/HumanCompatibleAI/KataGoVisualizer/pull/21/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

AdamGleave replied:
<pre class="ql-syntax" spellcheck="false">Line #68.                victim_df.groupby("adv_steps")
</pre><p>This group by is used in 3 different places so can probably just assign to a temp variable and reuse</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/HumanCompatibleAI/KataGoVisualizer/pull/21/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

AdamGleave replied:
<pre class="ql-syntax" spellcheck="false">Line #72.            df_confint = victim_df.groupby("adv_steps").mean(True)
</pre><p>confusing variable name? It's not really a confidence interval in its own right</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/HumanCompatibleAI/KataGoVisualizer/pull/21/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

AdamGleave replied:
<pre class="ql-syntax" spellcheck="false">Line #4.    victim_change_steps = [
</pre><p>Why is this hardcoded?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/HumanCompatibleAI/KataGoVisualizer/pull/21/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

AdamGleave commented on notebooks/notebooks/iclr2022/win-rate-vs-training-steps.ipynb:
```diff
@@ -0,0 +1,366 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #39.        ignored_adversaries=[],
</pre><p>In general having <code>[]</code> as a default argument is dangerous because it's a mutable object: if you change <code>ignored_adversaries</code> it'll get reflected in subsequent calls. I think it's harmless in this particular method, but I'd be inclined to change it to default to <code>None</code> just in case.   </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/HumanCompatibleAI/KataGoVisualizer/pull/21/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

tomtseng replied:
<p>I didn't know about this! fixed</p><p> </p>

AdamGleave commented on notebooks/notebooks/iclr2022/win-rate-vs-training-steps.ipynb:
```diff
@@ -0,0 +1,366 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #4.    victim_change_steps = [12560640, 16541696, 22263040, 25102336]
</pre><p>Why is this hardcoded?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/HumanCompatibleAI/KataGoVisualizer/pull/21/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

tomtseng replied:
<p>It takes a while to parse through all the training SGFs and extract victim_change_steps. But actually it doesn't take as much time as I thought initially, so I'll un-hardcode it</p><p> </p>

AdamGleave commented on notebooks/notebooks/iclr2022/win-rate-vs-visits.ipynb:
```diff
@@ -0,0 +1,460 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #1.    def plot_victim_visit_sweep(df, victim_label, plot_name):
</pre><p>type annotations?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/HumanCompatibleAI/KataGoVisualizer/pull/21/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

AdamGleave replied:
<pre class="ql-syntax" spellcheck="false">Line #2.        """Plot performance of A-MCTS-{R,S++}-v200 vs. varying victim visits."""
</pre><p>Overly specific docstring, will it only work with v200?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/HumanCompatibleAI/KataGoVisualizer/pull/21/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

AdamGleave replied:
<pre class="ql-syntax" spellcheck="false">Line #30.        # plt.title("Adversary against cp505")
</pre><p>delete?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/HumanCompatibleAI/KataGoVisualizer/pull/21/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

AdamGleave replied:
<pre class="ql-syntax" spellcheck="false">Line #38.        minor_ticks = utils.flatten_2d_list(
</pre><p>I found this line difficult to understand but it's actually doings something quite simple (add 9 ticks between each major tick in intervals of 10) -- add comment or edit to be more readily comprehensible?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/HumanCompatibleAI/KataGoVisualizer/pull/21/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

AdamGleave replied:
<pre class="ql-syntax" spellcheck="false">Line #44.        plt.savefig(f"{plot_name}.pgf", backend="pgf")
</pre><p>Could leave this as responsibility of callee and remove plot_name</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/HumanCompatibleAI/KataGoVisualizer/pull/21/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

AdamGleave replied:
<pre class="ql-syntax" spellcheck="false">Line #8.    plot_victim_visit_sweep(df, "$\\texttt{Latest}$", "adv-vs-cp505-vary-visits")
</pre><p>It's great this is all already factored into a helper method!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/HumanCompatibleAI/KataGoVisualizer/pull/21/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

AdamGleave replied:
<pre class="ql-syntax" spellcheck="false">Line #4.        df = data for adversary visit sweep experiment
</pre><p><code>df</code> and <code>df2</code> are not very informative names!  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/HumanCompatibleAI/KataGoVisualizer/pull/21/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

AdamGleave replied:
<pre class="ql-syntax" spellcheck="false">Line #23.        for name in df2_points.index:
</pre><p>Do we expect exactly two names in the index? If so, should we assert that (i.e. we're not overwriting existing values for splusplus/perfect_modeling points)? If not, is the index guaranteed to be sorted appropriately?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/HumanCompatibleAI/KataGoVisualizer/pull/21/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

AdamGleave replied:
<pre class="ql-syntax" spellcheck="false">Line #37.        (
</pre><p>Multi-line expression hard to read, consider introducing intermediate variable?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/HumanCompatibleAI/KataGoVisualizer/pull/21/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

AdamGleave reviewed (changes requested):
Thanks for working on this, a substantial improvement!

Some high-level comments:
  - The notebooks currently set formatting parameters by setting globals in `utils.set_plot_formatting()`, and then overriding some components such as figure size on a per-cell basis. I was wondering why you chose to do this rather than using matplotlib [style sheets and context managers](https://matplotlib.org/stable/tutorials/introductory/customizing.html) like `with plt.style.context(utils.STYLES['default'], utils.STYLES['1-col'])`. The main advantage I see of globals is you only need to set it once so it's less clutter. But it seems a bit error-prone, and potentially less flexible as some components of style depend on the figure size (e.g. we may want to adjust `figure.subplot.{top,left,bottom,right}` to have appropriately sized margins) and we don't have an easy way of coupling those together right now.
  - Do we really want the PGF files to be comittted to this repo? I'd be inclined to add them to `.gitignore` and instead just commit them in the `go-attack-paper` repo, but open to arguments to the contrary.

I've suggested a bunch of changes in the files, most notably refactoring duplicated code into helper methods. I'd suggest fixing the comments that are quick to do so, then just make a list of unaddressed comments and open an issue for them to address in future PR. Don't want to block this PR as splitting the notebooks up is clearly an improvement on status quo even if the code inside the notebooks remains messy.

tomtseng commented:
Addressed comments.

> The notebooks currently set formatting parameters by setting globals in utils.set_plot_formatting(), and then overriding some components such as figure size on a per-cell basis. I was wondering why you chose to do this rather than using matplotlib [style sheets and context managers](https://matplotlib.org/stable/tutorials/introductory/customizing.html) like with plt.style.context(utils.STYLES['default'], utils.STYLES['1-col']). 

I've added style sheets now. I didn't know about this feature of Matplotlib

> Do we really want the PGF files to be comittted to this repo? I'd be inclined to add them to .gitignore and instead just commit them in the go-attack-paper repo, but open to arguments to the contrary.

Maybe it's mildly useful to check whether notebook changes have changed the plots, and it's less likely that someone will accidentally upload old plots to go-attack-paper. But removing them seems fine — the plots are visible in the notebooks anyway

AdamGleave commented on notebooks/notebooks/iclr2022/utils.py:
```diff
@@ -0,0 +1,60 @@
+
+def get_style(style_name: str) -> str:
+    """Gets Matplotlib style sheet with a given name."""
+    return f"./matplotlib-style-sheets/{style_name}.mlpstyle"
```
Depending on CWD is a bit fragile, consider making it relative to the script path https://note.nkmk.me/en/python-script-file-path/ or https://docs.python.org/3/library/importlib.resources.html#module-importlib.resources

AdamGleave commented on streamlit_app/requirements.txt:
```diff
@@ -1,4 +1,4 @@
-./../go_attack_utils
+-e ./../go_attack_utils
```
Why do we need it to be develop mode here? Is there something wrong with the setup.py (missing packages, or package_data)?

Generally seems best to avoid -e outside of development. I feel OK with it for notebooks as you often *are* interactively developing the notebooks, but for a Dockerized app it feels a bit much.

tomtseng replied:
I think I was debugging something, added `-e` everywhere, and forgot to revert. Will remove

AdamGleave reviewed (approved):
LGTM, some minor suggested changes.

There's still a lot of refactoring that could be done but probably best to address in a separate PR. Consider making a ticket on the Kanban to track it and pointing to some of the unresolved comments in this PR so we don't forget about it?

tomtseng commented:
I think the only resolved comment is this one on `win-rate-vs-training-steps.ipynb`:

> The code in this particular notebook feels fundamental enough that we likely want to move it outside of a notebook into a Python module -- but that can be another PR.

I'll make a kanban task to clean up the notebooks more
